util/editEntry.js

```diff
@@ -0,0 +1,222 @@
+  }
+
+  try {
+    // Connect to the Database
+    var db = await connectToDb();
+  } catch (err) {
+    console.log("Unable to connect to MongoDB")
+    return { statusCode: 500, message: "Unable to connect to MongoDB Server", errorName: err.name, error: err.message }
+  }
+  // Check that all Id strings are Valid Mongo Object Ids
+  const validDevices = !data.summaryBOM.map((dev) => ObjectId.isValid(dev.deviceId)).includes(false);
+  const validTestCases = !data.testCases.map((str) => ObjectId.isValid(str)).includes(false);
+  const validObjectIds = validDevices && validTestCases && ObjectId.isValid(data.engagementId) && ObjectId.isValid(data._id);
+  if (validObjectIds) {
+    // TypeCast ID strings to Mongo ObjectId's
+    const id = ObjectId(validData._id);
+    const summaryBOM = validData.summaryBOM.map(device => {
+      return { ...device, deviceId: ObjectId(device.deviceId) }
+    });
+    const testCases = validData.testCases.map(testCaseId => ObjectId(testCaseId));
+    // Create the database query and replacement object
+    const query = { _id: id };
+    const newtestPlan = { ...validData, summaryBOM: summaryBOM, testCases: testCases, _id: id };
+    try {
+      // Update the Database with new TestPlan
+      var queryResult = await db.collection("testPlan").replaceOne(query, newtestPlan);
+    } catch (err) {
+      // Mongo-Side Validation failure should occur here
+      return { statusCode: 400, message: "MongoDB Query Failed or could not Validate", mongoQueryResult: queryResult, errorName: err.name, error: err.message }
+    }
+    // Edit was Successful!
+    return { statusCode: 200, message: "Success", mongoQueryResult: queryResult }
+  } else {
+    // Invalid Ids
+    return { statusCode: 422, message: "Validation Failed: Contains invalid MongoId(s)" }
+  }
+}
+
+export async function editTestCase(data) {
+  try {
+    // Validate Data
+    var validData = await testCaseSchema.validate(data, { abortEarly: false, stripUnknown: true });
+  } catch (err) {
+    return { statusCode: 422, message: "Yup Validation Failed", errors: err.errors }
+  }
+
+  try {
+    // Connect to the Database
+    var db = await connectToDb();
+  } catch (err) {
+    console.log("Unable to connect to MongoDB")
+    return { statusCode: 500, message: "Unable to connect to MongoDB Server", errorName: err.name, error: err.message }
+  }
+
+  // Check that all Id strings are Valid Mongo Object Ids
+  const validDevices = !data.BOM.map((dev) => ObjectId.isValid(dev.deviceId)).includes(false);
+  const validTests = !data.tests.map((str) => ObjectId.isValid(str)).includes(false);
+  const validObjectIds = validDevices && validTests && ObjectId.isValid(data.testPlanId) && ObjectId.isValid(data._id);
+
+  if (validObjectIds) {
+    // TypeCast ID strings to Mongo ObjectId's
+    const id = ObjectId(validData._id);
+    const testPlanId = ObjectId(validData.testPlanId);
+
+    const BOM = validData.BOM.map(device => {
```

Comment:
summaryBOM and BOM elements have been updated to consist of {_id, deviceId, isOptional, quantity}. So in addition to casting **deviceId**, should also cast **_id** as ObjectId.